### PR TITLE
CUDA static device memory

### PR DESCRIPTION
**Pull request by BenjaminW3**
This implements #209 (\__constant\__) and #60 (\__device\__).

### CODE REVIEW

File: test/unit/mem/view/src/ViewConstantMem.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        alpaka::mem::view::ViewPlainPtr<DevAcc, const Elem, Dim, Size> bufHost(data.data(), devAcc, extent);
+
+        auto viewConstantMemUninitialized(
+            alpaka::mem::view::createConstantMemView(
+                &g_constantMemory2DUninitialized[0u][0u],
+                devAcc,
+                extent));
+
+        alpaka::mem::view::copy(streamAcc, viewConstantMemUninitialized, bufHost, extent);
+
```
Comment (erikzenker) [inline test/unit/mem/view/src/ViewConstantMem.cpp]: I think there is an `alpaka::wait::wait(streamAcc)` missing here.

File: test/unit/mem/view/src/ViewConstantMem.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    //-----------------------------------------------------------------------------
+    // uninitialized global memory
+    {
+        using StreamAcc = alpaka::test::stream::DefaultStream<DevAcc>;
+        StreamAcc streamAcc(devAcc);
+
+        std::vector<Elem> const data{0, 1, 2, 3, 4, 5};
+        alpaka::mem::view::ViewPlainPtr<DevAcc, const Elem, Dim, Size> bufHost(data.data(), devAcc, extent);
```
Comment (erikzenker) [inline test/unit/mem/view/src/ViewConstantMem.cpp]: Doesn`t this needs to be a host buffer ?

File: test/unit/mem/view/src/ViewConstantMem.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        auto const gridThreadExtent =
+            alpaka::workdiv::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc);
+        auto const gridThreadIdx =
+            alpaka::idx::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+
+        auto const offset = gridThreadExtent[1u] * gridThreadIdx[0u] + gridThreadIdx[1u];
+        auto const val = offset;
+
+        BOOST_REQUIRE_EQUAL(
```
Comment (erikzenker) [inline test/unit/mem/view/src/ViewConstantMem.cpp]: Cuda does not like this macro, but I think that this is an known issue

File: test/unit/mem/view/src/ViewConstantMem.cpp (hunk 2 above)
Comment (BenjaminW3) [inline test/unit/mem/view/src/ViewConstantMem.cpp]: Yes, you are right.

File: test/unit/mem/view/src/ViewConstantMem.cpp (hunk 1 above)
Comment (BenjaminW3) [inline test/unit/mem/view/src/ViewConstantMem.cpp]: Yes, you are right.

**Comment by erikzenker** [comment]
Basically it works, but we have some problems:

* If we are using CPU and CUDA accelerator at the same time then all constant memory declarations are prefixed with `__constant__`, therefore this constant memory will live within the CUDA device. However, there is some [undocumented behaviour](http://stackoverflow.com/questions/34041372/access-cuda-global-device-variable-from-host) that for each constant memory on the device exist a constant memory on the host. This corresponding host memory is not initialized with the values from the initializer list (thats why the first test case will fail). Thus, the user has to initialize its constant memory by the copy all if it should work in all cases.

* Clang does not implement this undocumented behaviour mentioned above. Accordingly, no one know what happens when the CPU tries to access constant memory while GPU is also enabled.

**Comment by BenjaminW3** [comment]
I have fixed the bugs you noted and added support not only for `\__constant\__`, but also for `\__device\__` memory.
Could you please have a look at it again?

File: test/unit/mem/view/src/ViewConstantMem.cpp (hunk 3 above)
Comment (BenjaminW3) [inline test/unit/mem/view/src/ViewConstantMem.cpp]: I have added issue #230 to address this problem.

**Comment by erikzenker** [comment]
Thank you very much for that quick fix. The tests run through now (for cpu and gpu). I found no other issues while reading through the code. If there is no functionality left (WIP), then I think this is ready to merge.

**Comment by BenjaminW3** [comment]
I have removed the work-in-progress label.

**Pull request merged**